### ISSUE TITLE
ENH: Forward compatible "shape" argument vs "size"

### ISSUE THREAD

**Issue by jonas-eschle** [numpy/numpy#26916]
### Proposed new feature or change:

As briefly discussed at SciPy2024, forward compatible version of functions that take a "size" argument vs a "shape" would be welcome. Opening this here to ask for opinions before starting a PR.

# Idea

Some functions take a `shape` argument (`np.zeros`, as in the array API), others take a `size` (as np.random.... does), describing in their docs as `shape`. Other libraries like TensorFlow ([random.normal](https://www.tensorflow.org/api_docs/python/tf/random/uniform) take consistently a `shape` argument and the array API (afaik) is also tending towards this, albeit the randomness sector is hard to agree on (given the different generator designs) and no such formal decision has been made.
IMHO, `shape` will come and be consistent throughout in the array API and would anyway be the better (more consistent) choice.

# Proposal

Adding an additional `shape` keyword to functions that take a `size` as argument for forward compatibility.
If both are specified, the code errors. No other changes are intended internally (but open to if needed), it's meant to be a zero-cost PR.

## Goal

Provide a consistent naming as an input name for users. Getting this in ASAP means that less legacy code is being written, future upgrades will be easier and it's simply less burden to remember where to use `size` vs `shape` for users.

## No-goals

No deprecation or breaking changes are planned, no behavioral changes either. Simply use a consistent naming.

If there are special cases where `size` maybe fits better, this can be left untouched.

## Future

In the future, and with the API standard adapting this formally, `size` could be depricated, internals could be changed. But that's explicitly not part of the PR.

# Compatibility

This is meant as a forward compatibility feature and should not pose any problems. The only way this could backfire seems to be if another name would be adopted in the array API standard, this seems to me to unlikely to rule out. Even if it were, I would still argue that numpy should allow `shape`. But it seems like an unrealistic scenario.

# Actions

If people agree on this, I'll gladly open a PR and we'll move it there for the more detailed discussion.

**Comment by rkern** [comment]
As far as I know, it's only `np.random` that's the odd one out inside of numpy. This is not a widespread problem. 

One thing you will have to contend with inside of `np.random` are the methods that currently have a `shape` argument (e.g. for the shape parameter of a Gamma distribution) in addition to the `size`. I'm not sure how you would handle that in a zero-cost, forward-compatible manner.

My initial feeling is that numpy 2.0 was probably the window to get that change done.

In terms of a future Array API, I am [skeptical that it would or should happen](https://github.com/data-apis/array-api/issues/431#issuecomment-1150255274), but I could see something like a SPEC that at least standardizes on the general arguments for some subset of methods. And standardizing on `shape` rather than `size` for the desired shape of the output array could well be one of the things standardized over. I'm not terribly keen on running ahead of that speculative SPEC. It may also be that no one is actually interested in on standardizing on anything because what actually can be standardized over doesn't buy you a whole lot.

**Comment by seberg** [comment]
Yeah, agreed it should only be random functions (i.e. if there is something else, we should just fix it!).
Although random functions includes the SciPy stats random functions, and I suspect there are more `shape=` parameters there than in NumPy maybe.

> My initial feeling is that numpy 2.0 was probably the window to get that change done.

It would be helpful to see how big a change this is, and probably it is only worthwhile if SciPy random stats functions can also get away with it.

I don't know if we should do it, it is unclear enough to me that it should maybe have a (short!) NEP or SPEC just to make sure we have a rough idea of how, and how much, code would be broken (and improved).

The biggest problem is renaming existing `shape=` parameters.  It would be helpful to start that, because all of the steps can be done without a major release IMO.
But you may need to bring a lot of patience depending on impact and a major release can definitely help with the patience problem :) (which is a serious problem in practice).

But even with a major release, starting the process is helpful to smoothen the transition.  And if this is impactful to many users, starting the rename may even be required.
A major release allows to do more breaking things, but not to break many users in bad ways all at once.
(Not saying that this is a too big break: It has the potential to silently wrong results to just change the meaning, but plausibly the number of users affected that way would be very small.)

**Comment by jonas-eschle** [comment]
> As far as I know, it's only np.random that's the odd one out inside of numpy.

yes, agree!

> My initial feeling is that numpy 2.0 was probably the window to get that change done.

The discussion started in a "Numpy 2.0 feedback round", and this was my exact comment, that it would have been great to include this in 2.0 :) But just forward compatibility is probably fine (if it's just that)

> But even with a major release, starting the process is helpful to smoothen the transition. And if this is impactful to many users

Yes, this would be the main goal, just to get it started and easy any future transition, as it seems, it's a matter of _when_, not _if_.

> methods that currently have a shape argument

> The biggest problem is renaming existing shape= parameters.

I agree, that's probably the non-smooth part of it. I think two takes, one is as you proposed to go via NEP to back it and make the transition. Expecting headwinds, yes. My gut feeling would be that it's worth it though and needs to be like this, whether we like the way to get there or not. Or what's your estimate?

The other way as I envision it is a smaller PR: simply _adding_ shape whenever it doesn't collide. The argument is, that it _is_ already inconsistent in the API. So reducing the inconsistency to only a few dists that happen to have a shape parameter seems _better_, not the perfect solution, but better (the argument being that a user can always use shape, and the exceptions only occur if the specific parameter (that is mandatory and has some math meaning) is named `shape`, in which case using `size` makes sense). And completely backwards compatible.

In fact, it maybe highlights the inconsistency even more and adds some more arguments for why the `shape` (as a math parameter) to the dist should be changed.
Besides, shape seems like an inherent non-optimal name for a dist parameter, and the clash is in the nature of the existing name.

> Although random functions includes the SciPy stats random functions

True, since they're just moving to the new dists, they're maybe open to change it. Let's tap waters

Thanks a lot for thinking along here, @rkern @seberg I think getting a picture here of what might work and what doesn't

**Comment by charris** [comment]
Maybe a decorator in the short term?  We have done this before, although I don't recall which function/argument it was.

**Comment by seberg** [comment]
There are many things I would have liked in NumPy 2 that are not, but for the most part, I think it's a feature not a bug.  NumPy 2 is big enough, other releases will come, and few things are breaking enough that you cannot do them in a normal release anyway0.

---

In either case, to the interesting information:  Changing this in NumPy would be *very* easy as shape is only exclusively used for `gamma` and `standard_gamma` and it is the first argument so it is probably mostly passed positionally anyway.
(More importantly, it means that downstream can trivially be compatible to both versions by just passing it positionally.)

I suppose I am +0 for changing it in NumPy in that sense, but I don't have a strong opinion and I know the old `size=` choice was made for a good reason.  Did time how that the choice creates more confusion than it tries to avoid?  Maybe.
To me the main question might be if we mind diverging from `scipy.stats` matters.

Supporting both `shape=` and `size=` is obvious for a transition (maybe for years), but I would prefer not to do it forever.

> Maybe a decorator in the short term? We have done this before, although I don't recall which function/argument it was.

Yeah, I think Matt liberated something like this from SciPy, although it isn't too hard to implement either way.

**Comment by jonas-eschle** [comment]
> n either case, to the interesting information: Changing this in NumPy would be very easy as shape is only exclusively used for gamma and standard_gamma and it is the first argument so it is probably mostly passed positionally anyway.

This makes things indeed a lot simpler.

> and I know the old size= choice was made for a good reason.

Was it? Do you remember the reason? From a recent chat I remember that it was rather historical without a fundamental reason.
The crucial point seems to me that `size` and `shape` behave exactly the same (if I am correct).

> Supporting both shape= and size= is obvious for a transition (maybe for years), but I would prefer not to do it forever.

Fully agree.

> I suppose I am +0 for changing it in NumPy

I wonder why? This is a large pain point for anyone not being super familiar with numpy: one always has to look up the docs. A social test: asking people, how to generate an array in numpy, with `size` or `shape`, it's clear that the average user doesn't know and will look up the documentation (including myself here). There is no obvious need for this incosistency, and other libraries (like TF) have that nicely consistent (and it's a blessing).

Rethinking: if you were to design an array library from scratch and we already agreed (array API) that _some_ generating functions use `shape` as the argument. What would be the argument to say that we should change this name (although having identical behavior) in the random number generation?

> To me the main question might be if we mind diverging from scipy.stats matters.

I think to reformulate the proposal, it wouldn't because numpy _is_ already inconsistent in itself. So the question isn't, is numpy gonna be inconsistent, but _how_ much inconsistency do we have?

The strongest argument to me seems to be to _reduce_ the inconsistency. `shape` is the clear way forward, I would suggest to add this possibility where painless possible. And the two breaking changes could probably simply be solved (or even ignored: it's inconsistent, but _less_).

reformulated -> make numpy internally consisten by offering everywhere a shape argument where currently a `size` is accepted. No behavioral change.

> Maybe a decorator in the short term? We have done this before, although I don't recall which function/argument it was.

technically this seems like the elegant solution, maybe not politically, as this _could_ break things. `wraps` got most of the cases etc, but still, it's a different object. This _could_ pose the question of breaking compatibility (although it was never guaranteed to be the same _object_, AFAIU, users may still use it this way), and therefore may become more of a political hurdle.

All this being said, this seems to me convincing enough to open a PR and add the proposed changes, and then discuss details there.

**Comment by rkern** [comment]
> Was it? Do you remember the reason? From a recent chat I remember that it was rather historical without a fundamental reason.

To avoid confusion with the shape parameters of the various distributions. This was first done in `scipy.stats` back in the super-early days when it was running on Numeric. While mostly not explicitly conflicting with method arguments named `shape`, they do talk about all of the parameters that are not `loc` and `scale` as "shape parameters" (as is standard mathematical practice) and did not want the confusion with having `loc=0, scale=1, shape=None` where `shape` actually meant the shape of the desired array. When I wrote [`mtrand`](https://pypi.org/project/mtrand/), I followed suit so it could be a drop-in replacement for the PRNG code that was in there. Then numpy adopted `mtrand` as `numpy.random`. IIRC, this was while we were still considering it to be `scipy_core`, so `scipy.stats` having made that decision to be somewhat inconsistent with the rest, but a fairly self-contained and *self*-consistent inconsistency withing the one region made some sense.

[I am keenly, *keenly* aware of the tabs in `Makefile`s story.](https://beebo.org/haycorn/2015-04-20_tabs-and-makefiles.html)

> This is a large pain point for anyone not being super familiar with numpy: one always has to look up the docs.

I'd love to see an actual experiment that demonstrates that.

**Comment by jonas-eschle** [comment]
> To avoid confusion with the shape parameters of the various distributions.

Thanks a lot for the whole explanation, I wasn't aware of this. And yes, one should be careful when naming things, no question.

> shape parameters

I agree, but `shape` is not `shape parameters` or `parameters`. I think `shape` itself is quite established to describe the shape of the array (see also below), and I don't really see any use-case to use `shape` to describe an actual parameter in a mathematical distribution? (apart of the gamma, but there is no mathematical need for it).

So I see where this may comes from, and the foresight was surely a good thing, but as times passes, `shape` has established itself as a clear keyword in the realm of arrays across libraries I think.

Another consistency point is that the shape attribute pretty much aligns with the `shape` attribute of an array, but not with the `size` of it.
Moreover, just to emphasize, the docstring of `size` says that this defines the shape.

So not naming the shape `shape` for pre-caution of clashing: I see that! But there seems no actual clash to be happening now (which is easy in hindsight of course, but it's what we've learned over time) and we can make the naming more consistent. Or would you disagree with this?

> I'd love to see an actual experiment that demonstrates that

In my experience, any numpy user that I ask (and isn't a super power user) "how do you create an array in numpy, with `shape` or `size`" (I could count dozens of examples with zero exceptions, just being at SciPy conf) will not have an answer and be confused, confessing that he/she has to look it up every time and always forget. Or did you _ever_ had someone say the opposite (who isn't a power user)? I think the _pain_ of this is probably only realized once one uses a library that is consistent (i.e. TensorFlow) and then comes back to numpy.

**Comment by rkern** [comment]
> So not naming the shape `shape` for pre-caution of clashing: I see that! But there seems no actual clash to be happening now (which is easy in hindsight of course, but it's what we've learned over time) and we can make the naming more consistent. Or would you disagree with this?

It was not about clashing with another parameter named exactly `shape`, per se. It was about ensuring that people understood that that parameter was not a shape parameter of the distribution.

I am not going to argue that it was an ideal choice; I was not the one who made it. But I will assure you that the inconsistency was done deliberately, knowing full well how `shape` was used elsewhere (said person being the main author of `numpy` as well as the distributions in `scipy.stats`).

> In my experience, any numpy user that I ask (and isn't a super power user) "how do you create an array in numpy, with `shape` or `size`" (I could count dozens of examples with zero exceptions, just being at SciPy conf) will not have an answer and be confused, confessing that he/she has to look it up every time and always forget.

I have my suspicions that `np.random` is not causally related to this. It's a pretty confusing question in its own right given that most places where `shape` is used "to create an array in `numpy`" (`ones()`, `zeros()`, `full()`) usually get that argument passed positionally. If you want to run this experiment for real, have people write code (in an environment with controlled access to docs) where they'd pass said argument as a keyword (the `np.random` methods, or things like `fromstring()`, I guess). I'm willing to be most of the errors will be people using `shape=` for `np.random` methods rather than a general overall confusion.

> I think the _pain_ of this is probably only realized once one uses a library that is consistent (i.e. TensorFlow) and then comes back to numpy.

To be honest, I will stack up the pains I feel every time I use a creaking, old API I designed a couple of decades ago against anyone.

**Comment by seberg** [comment]
To be clear I am +0 as in: I don't feel like pushing it especially without understanding the history first (Chesterton's fence).  But since Robert for one doesn't seem attached to avoiding shape in random (if I understand that correctly).  Maybe a good chance to nobody really prefers the `size` anymore.

FWIW, I agree that "shape" should already be the clear choice for practically anyone, so some might try it on random and be briefly surprised mainly.  (which is a fair point)

**Comment by rkern** [comment]
I'm -0 on basic disruption reasons, which I feel more strongly about than my regret for propagating `size`.

I don't agree that the middle state, where both `shape=` and `size=` are accepted, is strictly better than the status quo. I think such states are necessary evils if the end state, where the one desired name is accepted, is being actively moved towards, to give time for people to implement the transition. I think having it be an endless limbo is going to be _more_ confusing for users than the current state. And the alternative, where we actively pursue deprecation and removal, is more disruptive than (I think) it's worth (or else I would have pushed for it long ago), given the API surface that it touches. This is a lot of methods, not just a handful of functions. Including the `RandomState` methods that we've promised to keep frozen. This is why I think numpy 2.0 was the window to get something like this done.

**Comment by jonas-eschle** [comment]
> It was not about clashing with another parameter named exactly `shape`, per se. It was about ensuring that people understood that that parameter was not a shape parameter of the distribution.

Makes sense! I would argue that today, nobody would misunderstand that, given the prevalence of `shape` as attributes and documentation all over libraries as specifying the "array-shape".

> said person being the main author of numpy as well as the distributions in scipy.stats

Who exactly is said person? Any possibility to include them here? But even then, although he may has done tremendous work and many good design decisions, this doesn't mean that things cannot, with the benefit of hindsight, be further improved.

> But I will assure you that the inconsistency was done deliberately, knowing full well how `shape` was used elsewhere (said person being the main author of `numpy` as well as the distributions in `scipy.stats`).

Thanks, it's a good reminder. As far as I can tell though, it was historical, and the API just needed to be invented. In hindsight it's easy to say that a consistent choice would have been better, but that's how APIs are designed (and more importantly, evolve, as we learn more).

> I'm willing to be most of the errors will be people using `shape=` for `np.random` methods rather than a general overall confusion.

I agree, or the positional only: just asked the shape vs size arg some colleagues the other day, and the answer was, it's positional (which it isn't in `np.random`).

> To be honest, I will stack up the pains I feel every time I use a creaking, old API I designed a couple of decades ago against anyone.

it's just normal, isn't it :)
But we cannot use the old justifications with the new knowledge.

> basic disruption reasons, which I feel more strongly about than my regret for propagating size.

> I don't agree that the middle state, where both shape= and size= are accepted, is strictly better than the status quo.

I think this has to be broken down into the two components:
- we want to avoid more confusing legacy code to be written (i.e. that `shape` is used everywhere). This is the most crucial point to me. Every single day, new people start using numpy, or I am even teaching it myself to students. This is the point about it: it's a minor thing, but multiplied by so many users!
- don't break anyones code/force them to change things (or as little as possible)

Simply introducing in a "forward compatible", zeroish breaking release the `shape` arguments will allow the first point to work. If the docs are clear, I don't expect a lot of confusion, or rather, the main confusion being about what `size` means and why one can use `size` too?

From there, the docs can be adjusted, shape can be moved up in the docs, we can deprecate in the docs, tutorials can be changed etc. And code deprecations _can_ be added. But I see it as a soft steering. The overhead of answering questions like "why is there also a `size` agrument to the `shape`" with "historic, just use `shape` everywhere" seems to me smaller or equal to the overhead of the confusion about what to use. Once you get the reason, the usage of `shape` over `size` is straightforward and doesn't require anymore thinking.

But I agree, that there should be a push towards `shape`; and I think nobody really disputes this. Yes, there would need to be deprecations at some point (abeit it being a fairly trivial change overall and a descriptive deprecation could help a lot). And at some point, it should be breaking, but this can be a period of 5+ years in my mind at least (long, very, very long). 


> Including the RandomState methods that we've promised to keep frozen.

"Practicality beats purity.". It's great to make promises that others can rely on. But then we're living in reality. And I don't think someone will really question this desicion.

> This is why I think numpy 2.0 was the window to get something like this done.

Fully agree, that was my initial motivation. "Now is better than never." ;)
